### PR TITLE
USAGOV-354-Update-Node-Modules

### PR DESCRIPTION
**Pull request by cwacht**
Switch from uswds-gulp to uswds-compile and switch node from alpine to bullseye

## Jira Task

https://cm-jira.usa.gov/browse/USAGOV-354

## Checklist for the Developer

- [ ] A link to the JIRA ticket has been added above.
- [ ] The JIRA ticket identifies the desired result of this change.
- [ ] The JIRA ticket contains clear acceptance criteria.
- [ ] The JIRA ticket contains clear testing steps.
- [ ] The JIRA ticket contains a description of "Done".
- [ ] Any preparation/installation/update steps required for this code to work properly (drush commands, scripts, configuration updates) are documented in the ticket.

## Checklist for the Peer Reviewers
- [ ] The branch name of this PR matches the project standards.
- [ ] QA steps are followed and any changes to process are updated in the ticket.
- [ ] Code Standards are followed, there are no bad practices in use.
- [ ] Config changes (if any) include only necessary changes.
- [ ] No errors in Drupal or Client Browser.
- [ ] Code has been tested locally (if possible).
- [ ] Following AC and Testing Steps verify changes work as expected.
- [ ] There are no known side-effects outside of expected behavior.

### CODE REVIEW

**Comment by akf** [comment]
I intend to review this, soon! We'll need to either dial down the javascript linting or fix some things in our js -- or both. Probably the same applies for css.

**Comment by IsabelLaurenceau** [comment]
Screen Shot 2023-01-30 at 10 54 34 AM
Screen Shot 2023-01-30 at 10 54 27 AM
Screen Shot 2023-01-30 at 10 54 20 AM

Looks like something was off with the styling when I made a new build from this branch this morning.

I didn't get any error messages in the terminal; the logs looked similar to other builds I do

**Comment by akf** [comment]
I'm seeing pages similar to what @IsabelLaurenceau does. The USWDS compilation generates a lot fewer messages than the old version, which probably doesn't mean anything.

**Comment by cwacht** [comment]
With the 2 commits that I made today, I think I've gotten my local linting output to match what I see on Circle CI.
Question, do I need to fix all the linting errors before we can approve this pull request? If so, should I fix them in this pull request?

**Comment by akf** [comment]
I think they should all be fixed (or the linting rules tweaked further -- it looks like we can downgrade specific rules from "error" pretty easily in eslint, I didn't look at stylelint). Basically, the linter needs to pass; otherwise we'll just learn to ignore it. 
That doesn't mean you need to do all the work! We can split this up. If you want to do that, it's probably best if people make branches off of this one and you merge them in. I'll sign up for at least some of the js work. 

I see this at the end of the report: 

"125 errors and 0 warnings potentially fixable with the `--fix` option." 

Maybe running "npx eslint --fix scripts/*" would be a good first step?

File: .docker/Dockerfile-cms
Comment (akf) [inline .docker/Dockerfile-cms]: Let's delete the commented-out code.

**Comment by akf** [comment]
Just FYI: I need to look into this further ... 

This still has the problem where a fresh build (in my case, using the pr-review script, which is a bit suspect) winds up missing the styles -- more of them than last time! "git diff" shows me this: 

	modified:   web/themes/custom/usagov/package-lock.json
	modified:   web/themes/custom/usagov/sass/_uswds-theme-custom-styles.scss
	modified:   web/themes/custom/usagov/sass/_uswds-theme.scss

It looks like both of the scss files have been reverted to what a fresh USWDS build would have, probably. 

I'm looking at my latest notes on the Jira ticket and not quite following something I wrote about how we're invoking "npm init," and I'm not tracking whether / where we're invoking that now.

**Comment by cwacht** [comment]
My fix to prevent those files from being overridden was to change `bin/npm run init` to `bin/npm run build` in the /bin/init file.

My best guess is that if you were running the pr-review script while on any other branch, then your /bin/init file probably still ran `bin/npm run init` which would cause those files to get overridden.

**Comment by akf** [comment]
I'm probably confused about something -- bin/init on the dev branch doesn't have a call to "bin/npm run init" (anymore?) that I can see. 
I'm going to try this from a fresh clone of the repo, though.

**Comment by cwacht** [comment]
It looks like you're right about that. I dont know what else would have overwritten those files. Make sure that those changes are not present when you start a fresh build.

**Comment by akf** [comment]
On a totally fresh installation (cleared out all my docker images, even), I'm getting a different error and my build doesn't complete -- IIRC it failed to build the node image as well as a couple of other things. The error message mentions making sure I have "sass-embedded-linux-arm64" installed, which makes me think this might be an issue that hits M1 Macs only. Here's the error, plus a couple of lines above. I'm going to try to figure out what I might need to add here, but I thought I'd capture it: 

```
#0 11.74 > usagov_sass@1.0.0 build
#0 11.74 > npx gulp compile
#0 11.74 
#0 12.40 Error: Embedded Dart Sass couldn't find the embedded compiler executable. Please make sure the optional dependency sass-embedded-linux-arm64 is installed in node_modules.
#0 12.40     at /var/www/web/themes/custom/usagov/node_modules/sass-embedded/dist/lib/src/compiler-path.js:28:11
#0 12.40     at Object.<anonymous> (/var/www/web/themes/custom/usagov/node_modules/sass-embedded/dist/lib/src/compiler-path.js:32:3)
#0 12.40     at Module._compile (node:internal/modules/cjs/loader:1218:14)
#0 12.40     at Module._extensions..js (node:internal/modules/cjs/loader:1272:10)
#0 12.40     at Module.load (node:internal/modules/cjs/loader:1081:32)
#0 12.40     at Module._load (node:internal/modules/cjs/loader:922:12)
#0 12.40     at Module.require (node:internal/modules/cjs/loader:1105:19)
#0 12.40     at require (node:internal/modules/cjs/helpers:103:18)
#0 12.40     at Object.<anonymous> (/var/www/web/themes/custom/usagov/node_modules/sass-embedded/dist/lib/src/async-compiler.js:10:25)
#0 12.40     at Module._compile (node:internal/modules/cjs/loader:1218:14)
```

**Comment by akf** [comment]
Confirmed -- this is an M1 Mac vs. Intel issue. I was able to get up and running by removing the package-lock.json file and running bin/init again; this time the node build pulled in arm64-specific dependencies. No other files were modified! 

I think we want the package-lock file, though. So I'm noodling on how best to handle this.

File: .docker/Dockerfile-cms
Comment (akf) [inline .docker/Dockerfile-cms]: Change 19.3 to 19.9 (newer version with fewer vulnerabilities)

**Comment by akf** [comment]
Apparently there is an npm bug (or perhaps there's disagreement over whether it's a bug) that results in different package-lock.json contents depending on whether you already have node_modules installed when you install or update packages. I've created a pull request to merge into _this branch_ and tagged you as a reviewer -- please see if that works on your machine, and if so, merge it in.

**Review by akf — CHANGES_REQUESTED** [review]
Question: did you add "use strict" to the javascript functions because eslint seemed to be asking for it? I think we should make sure to test all the modified javascript just in case. 

Beyond that, I think I'm done reviewing. If my change to the package-lock.json file works for you, I would approve this with that PR merged in. Would like to aim to merge this shortly after a deployment to Staging (so, on a Friday) to allow for regression testing on dev.

**Review by akf — CHANGES_REQUESTED** [review]
I've made another pull request for a change that fixed the missing files problem on my local dev site: https://github.com/usagov/usagov-2021/pull/893. (Doing the PR partly to make it easy to track the changes, but also because I think if I commit to this branch I can't approve it anymore!)

I don't know why this would have worked for some builds, and my best guess is that we weren't really starting from scratch when we had "success" before. 

I recommend @IsabelLaurenceau  as a third reviewer. Another test deployment to the dev space will definitely be a good review step as well.

**Comment by cwacht** [comment]
OK I just pushed some changes that do a different fix, but I like your way better, so I will revert my changes and merge yours in. I'm honestly confused about why your solution is not the default starting point for USWDS compile...

**Review by akf — APPROVED** [review]
Now working on your machine, my machine, and dev!

**Pull request merged**
